### PR TITLE
[#804] feat(mysql): Support mysql table properties.

### PR DESCRIPTION
**Pull request by Clearvive**
### What changes were proposed in this pull request?

Support mysql table properties.

### Why are the changes needed?
Fix: #804 

### Does this PR introduce _any_ user-facing change?
Add mysql properties support.

### How was this patch tested?
UT/IT

### CODE REVIEW

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/JdbcCatalog.java [hunk 1]
```diff
@@ -63,4 +66,18 @@ protected JdbcExceptionConverter createExceptionConverter() {
 
   /** @return The {@link JdbcTableOperations} to be used by the catalog to manage tables in the */
   protected abstract JdbcTableOperations createJdbcTableOperations();
+
+  /**
+   * @return The {@link JdbcTablePropertiesMetadata} to be used by the catalog to manage tables in
+   *     the
```
Comment (mchades) [inline catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/JdbcCatalog.java:72]: Incomplete sentence?

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/MysqlTablePropertiesMetadata.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+  static {
+    PROPERTIES_METADATA =
+        Collections.unmodifiableMap(
+            new HashMap<String, PropertyEntry<?>>() {
+              {
+                put(COMMENT, stringReservedPropertyEntry(COMMENT, "The table comment", true));
+                put(
+                    ENGINE,
+                    stringOptionalPropertyEntry(ENGINE, "The table engine", false, null, false));
```
Comment (mchades) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/MysqlTablePropertiesMetadata.java:32]: Wouldn't it be better to define it as an enum type?

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/MysqlTablePropertiesMetadata.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+                put(
+                    DEFAULT_CHARSET,
+                    stringOptionalPropertyEntry(
```
Comment (mchades) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/MysqlTablePropertiesMetadata.java:35]: ditto

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/MysqlTablePropertiesMetadata.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import com.datastrato.gravitino.catalog.PropertyEntry;
+import com.datastrato.gravitino.catalog.jdbc.JdbcTablePropertiesMetadata;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+public class MysqlTablePropertiesMetadata extends JdbcTablePropertiesMetadata {
+  public static final String ENGINE = "ENGINE";
+  public static final String DEFAULT_CHARSET = "DEFAULT CHARSET";
+  public static final String COLLATE = "COLLATE";
```
Comment (yuqi1129) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/MysqlTablePropertiesMetadata.java:20]: Currently we only support this table properties?

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/MysqlTablePropertiesMetadata.java (hunk 3 above)
Comment (yuqi1129) [reply catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/MysqlTablePropertiesMetadata.java:35]: If the value of them has little to do with version changes, I suggest using enumeration as we can verify its validity.

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/MysqlTablePropertiesMetadata.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    NDBCLUSTER("ndbcluster"),
+    FEDERATED("FEDERATED"),
+    MEMORY("MEMORY"),
+    INNODB("InnoDB"),
+    PERFORMANCE_SCHEMA("PERFORMANCE_SCHEMA"),
+    MYISAM("MyISAM"),
+    NDBINFO("ndbinfo"),
+    MRG_MYISAM("MRG_MYISAM"),
+    BLACKHOLE("BLACKHOLE"),
+    CSV("CSV"),
```
Comment (yuqi1129) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/MysqlTablePropertiesMetadata.java:69]: Is the engine name case-insensitive? If so, could you uniform the case here?

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/mysql/TestMysqlTableOperations.java [hunk 6]
```diff
@@ -77,9 +80,10 @@ public void testOperationTable() {
             .withNullable(false)
             .build());
     Map<String, String> properties = new HashMap<>();
-    // TODO #804 Properties will be unified in the future.
-    //    properties.put("ENGINE", "InnoDB");
-    //    properties.put(AUTO_INCREMENT, "10");
+    properties.put(ENGINE_KEY, "InnoDB");
+    properties.put(AUTO_INCREMENT, "10");
+    properties.put(DEFAULT_CHARSET_KEY, "utf8mb3");
+    properties.put(COLLATE_KEY, "utf8mb3_general_ci");
```
Comment (mchades) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/mysql/TestMysqlTableOperations.java:86]: Did you also need to test the default properties

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/MysqlTablePropertiesMetadata.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+                put(
+                    COLLATE_KEY,
+                    stringOptionalPropertyEntry(
+                        COLLATE_KEY, "The table collate", false, null, false));
+                // auto_increment properties can only be specified when creating and cannot be
+                // modified.
+                put(
+                    AUTO_INCREMENT_KEY,
+                    integerOptionalPropertyEntry(
+                        AUTO_INCREMENT_KEY, "The table auto increment offset", true, null, false));
```
Comment (mchades) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/MysqlTablePropertiesMetadata.java:54]: As the comment described, I think it should be a reserved property

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/JdbcTablePropertiesMetadata.java [hunk 8]
```diff
@@ -5,14 +5,8 @@
...
-import java.util.Map;
 
-public class JdbcTablePropertiesMetadata extends BasePropertiesMetadata {
+public abstract class JdbcTablePropertiesMetadata extends BasePropertiesMetadata {
 
-  @Override
-  protected Map<String, PropertyEntry<?>> specificPropertyEntries() {
-    return Collections.emptyMap();
-  }
+  public static final String COMMENT = "comment";
```
Comment (mchades) [inline catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/JdbcTablePropertiesMetadata.java:11]: Where is the metadata of the `COMMENT` property?

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/mysql/TestMysqlTableOperations.java [hunk 9]
```diff
@@ -77,9 +80,10 @@ public void testOperationTable() {
             .withNullable(false)
             .build());
     Map<String, String> properties = new HashMap<>();
-    // TODO #804 Properties will be unified in the future.
-    //    properties.put("ENGINE", "InnoDB");
-    //    properties.put(AUTO_INCREMENT, "10");
+    properties.put(ENGINE_KEY, "InnoDB");
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/mysql/TestMysqlTableOperations.java:83]: should we test other properties not managed by Gravitino?

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/JdbcTablePropertiesMetadata.java (hunk 8 above)
Comment (mchades) [reply catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/JdbcTablePropertiesMetadata.java:11]: So you should define it as a reserved property, right?

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/MysqlTablePropertiesMetadata.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+import com.datastrato.gravitino.catalog.PropertyEntry;
+import com.datastrato.gravitino.catalog.jdbc.JdbcTablePropertiesMetadata;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+public class MysqlTablePropertiesMetadata extends JdbcTablePropertiesMetadata {
+  public static final String ENGINE_KEY = "ENGINE";
+  public static final String DEFAULT_CHARSET_KEY = "DEFAULT CHARSET";
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/MysqlTablePropertiesMetadata.java:20]: I remeber the property name managed by Gravitino should use '-' to combine words, like `default-charset`

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/mysql/TestMysqlTableOperations.java (hunk 6 above)
Comment (mchades) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/mysql/TestMysqlTableOperations.java:86]: But the test image version we used is fixed. I suggest the scenario that needs testing is whether these properties are indeed automatically created by the underlying system.

**Comment by FANNG1** [comment]
could you explain the reason why add `ENGINE`, `DEFAULT CHARSET`, `COLLATE ` and `AUTO_INCREMENT ` to Mysql table properties meta to be managed by Gravitino.

**Comment by FANNG1** [comment]
do you support change table properies or remove table properties ?

**Comment by FANNG1** [comment]
> > could you explain the reason why add `ENGINE`, `DEFAULT CHARSET`, `COLLATE ` and `AUTO_INCREMENT ` to Mysql table properties meta to be managed by Gravitino.
> 
> It's solely because these four attributes are common and frequently used values in production.

I don't see a strong reason, especially for `DEFAULT CHARSET` `COLLATE`

**Comment by FANNG1** [comment]
> > e
> 
> image Not supported

seems we should support change table properites ?

**Comment by FANNG1** [comment]
I failed to retrive the properies from JDBC drivers, but instead of querying data from internal system table, I prefer to use `SHOW TABLE STATUS WHERE Name = 'xx'` to get the properties, seems more maintanable.

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java [hunk 11]
```diff
@@ -78,26 +79,57 @@ public JdbcTable load(String databaseName, String tableName) throws NoSuchTableE
...
       return new JdbcTable.Builder()
           .withName(tableName)
           .withColumns(jdbcColumns.toArray(new JdbcColumn[0]))
           .withComment(comment)
-          .withProperties(Collections.emptyMap())
+          .withProperties(
+              Collections.unmodifiableMap(
+                  new HashMap<String, String>() {
+                    {
+                      put(MYSQL_ENGINE_KEY, tableProperties.get(MYSQL_ENGINE_KEY));
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/operation/MysqlTableOperations.java:100]: is it possible tableProperties.get(MYSQL_ENGINE_KEY) returns null?

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/mysql/TestMysqlTableOperations.java [hunk 12]
```diff
@@ -462,12 +479,6 @@ public void testCreateAllTypeTable() {
             .withType(Types.BinaryType.get())
             .withNullable(false)
             .build());
-    columns.add(
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/mysql/TestMysqlTableOperations.java:465]: why remove this?

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/MysqlTablePropertiesMetadata.java [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+import com.datastrato.gravitino.catalog.PropertyEntry;
+import com.datastrato.gravitino.catalog.jdbc.JdbcTablePropertiesMetadata;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+public class MysqlTablePropertiesMetadata extends JdbcTablePropertiesMetadata {
+  public static final String GRAVITINO_ENGINE_KEY = "engine";
+  public static final String MYSQL_ENGINE_KEY = "ENGINE";
+  public static final String GRAVITINO_AUTO_INCREMENT_OFFSET_KEY = "auto-increment-offset";
```
Comment (yuqi1129) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/MysqlTablePropertiesMetadata.java:20]: What's the meanig of `offset`?

File: catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/MysqlTablePropertiesMetadata.java [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+  public static final Map<String, String> GRAVITINO_CONFIG_TO_MYSQL =
+      Collections.unmodifiableMap(
+          new HashMap<String, String>() {
+            {
+              put(GRAVITINO_ENGINE_KEY, MYSQL_ENGINE_KEY);
+              put(GRAVITINO_AUTO_INCREMENT_OFFSET_KEY, MYSQL_AUTO_INCREMENT_OFFSET_KEY);
+            }
+          });
+
+  public static final Map<String, String> MYSQL_CONFIG_TO_GRAVITINO =
```
Comment (yuqi1129) [inline catalogs/catalog-jdbc-mysql/src/main/java/com/datastrato/gravitino/catalog/mysql/MysqlTablePropertiesMetadata.java:33]: I guess you can use 

```java
  public static final BidiMap<String, String> map = new TreeBidiMap<>() {
    {
      put(GRAVITINO_ENGINE_KEY, MYSQL_ENGINE_KEY);
      put(GRAVITINO_AUTO_INCREMENT_OFFSET_KEY, MYSQL_AUTO_INCREMENT_OFFSET_KEY);
    }
  };

    map.inverseBidiMap().get("xxxx");
    map.get("xxxxx");
```

**Comment by yuqi1129** [comment]
It's okay for me and have not further comments.

**Review by yuqi1129 — APPROVED** [review]
LGTM

**Comment by jerryshao** [comment]
@FANNG1 @yuqi1129  you can go ahead if you feel OK.

Besides, we need to also update the doc @Clearvive

**Pull request merged**